modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/PagesList.java

```diff
@@ -1636,6 +1864,197 @@ private void changed() {
+        private final Object[] stripeLocks = new Object[STRIPES_COUNT];
+
+        /** Page lists. */
+        private final GridLongList[] stripes = new GridLongList[STRIPES_COUNT];
+
+        /** Atomic updater for nextStripeIdx field. */
+        private static final AtomicIntegerFieldUpdater<PagesCache> nextStripeUpdater = AtomicIntegerFieldUpdater
+            .newUpdater(PagesCache.class, "nextStripeIdx");
+
+        /** Atomic updater for size field. */
+        private static final AtomicIntegerFieldUpdater<PagesCache> sizeUpdater = AtomicIntegerFieldUpdater
+            .newUpdater(PagesCache.class, "size");
+
+        /** Access counter to provide round-robin stripes polling. */
+        private volatile int nextStripeIdx;
+
+        /** Cache size. */
+        private volatile int size;
+
+        /** Count of flush calls with empty cache. */
+        private int emptyFlushCnt;
+
+        /**
+         * Default constructor.
+         */
+        public PagesCache() {
+            assert U.isPow2(STRIPES_COUNT) : STRIPES_COUNT;
+
+            for (int i = 0; i < STRIPES_COUNT; i++)
+                stripeLocks[i] = new Object();
+        }
+
+        /**
+         * Remove page from the list.
+         *
+         * @param pageId Page id.
+         * @return {@code True} if page was found and succesfully removed, {@code false} if page not found.
+         */
+        public boolean removePage(long pageId) {
+            int stripeIdx = (int)pageId & (STRIPES_COUNT - 1);
+
+            synchronized (stripeLocks[stripeIdx]) {
+                GridLongList stripe = stripes[stripeIdx];
+
+                boolean rmvd = stripe != null && stripe.removeValue(0, pageId) >= 0;
+
+                if (rmvd)
+                    sizeUpdater.decrementAndGet(this);
+
+                return rmvd;
+            }
+        }
+
+        /**
+         * Poll next page from the list.
+         *
+         * @return pageId.
+         */
+        public long poll() {
+            if (size == 0)
+                return 0L;
+
+            for (int i = 0; i < STRIPES_COUNT; i++) {
+                int stripeIdx = nextStripeUpdater.getAndIncrement(this) & (STRIPES_COUNT - 1);
+
+                synchronized (stripeLocks[stripeIdx]) {
+                    GridLongList stripe = stripes[stripeIdx];
+
+                    if (stripe != null && !stripe.isEmpty()) {
+                        sizeUpdater.decrementAndGet(this);
+
+                        return stripe.remove();
+                    }
+                }
+            }
+
+            return 0L;
+        }
+
+        /**
+         * Flush all stripes to one list and clear stripes.
+         */
+        @SuppressWarnings("NonAtomicOperationOnVolatileField")
+        public GridLongList flush() {
+            GridLongList res = null;
+
+            if (size == 0) {
+                boolean stripesChanged = false;
+
+                if (++emptyFlushCnt >= EMPTY_FLUSH_GC_THRESHOLD) {
```

Comment:
emptyFlushCnt is not volatile, please add a comment why. I guess that it can be changed only from a checkpoint thread, but this might not be obvious later.

Reply:
yep, all calls under exclusive lock or in case of destroyCacheDataStore0 will be called without conc, is it enough or additional description needed ?

Reply:
It's enough I think